mlrun/model.py

```diff
@@ -1490,14 +1490,35 @@ def error(self) -> str:
             )
         return ""
 
-    def output(self, key):
-        """return the value of a specific result or artifact by key"""
+    def output(self, key: str):
+        """
+        Return the value of a specific result or artifact by key.
+
+        This method waits for the outputs to complete and retrieves the value corresponding to the provided key.
+        If the key exists in the results, it returns the corresponding result value.
+        If not found in results, it attempts to fetch the artifact by key.
+        If the artifact is not found, it tries to fetch the artifact URI by key.
+        If no artifact or result is found for the key, returns None.
+
+        :param key: The key of the result or artifact to retrieve.
+        :return: The value of the result or the artifact URI corresponding to the key, or None if not found.
+        """
         self._outputs_wait_for_completion()
+
+        # Check if the key exists in results and return the result value
         if self.status.results and key in self.status.results:
-            return self.status.results.get(key)
+            return self.status.results[key]
+
+        # Fetch the artifact for the key if available
+        # There are cases when `status.artifact_uri` does not exist in the run (e.g., when running locally).
         artifact = self._artifact(key)
         if artifact:
             return get_artifact_target(artifact, self.metadata.project)
+
+        # There are cases when `status.artifacts` does not exist in the run (e.g., when listing runs).
+        if self.status.artifact_uris and key in self.status.artifact_uris:
+            return self.status.artifact_uris[key]
+
```

Comment:
The logic is good but I think its better to have one concentrated comment instead of giving bits of information every few lines. Instead of these comments, write something along the lines of:
```
Artifacts are usually cached in the run object under `status.artifacts`. However, the artifacts are not stored in the DB as part of the run. The server may enrich the run with the artifacts or provide `status.artifact_uris` instead. See mlrun.common.formatters.run.RunFormat.

```